### DISCUSSION TITLE
What is the current best way to build an iOS app on Windows?

### DISCUSSION THREAD

**Discussion by mrfrasier** [airsdk/Adobe-Runtime-Support#3121] [Q&A]
Since late last year the AIR SDK has been plagued with issues affecting submissions to Apple. First builds crashed on iOS versions older than 16, and more recently Apple is rejecting builds for using APIs.

I've seen mentions of fixes coming soon and workarounds using an external mac to build etc, but I haven't seen any writeups or definitive answers and get no responses when I respond to various posts (here and other githubs related to AIR).

I can't be alone here. AIR devs using Windows, how are you currently submitting apps to Apple?

**Comment by FliplineStudios** [comment]
Personally I switched to Mac for packaging our iOS apps to avoid all of the issues you mentioned, it seemed too risky to package on Windows and have unpredictable issues on older iOS versions.

In the past I would normally package from the command line for Windows using a .bat script, so instead I just wrote a similar shell script on the Mac that packages it the same way.  I still do all of my development on Windows, I just have to copy the files over to the Mac and double-click a script there instead to package it up.

**Reply by mrfrasier** [reply]
Seems like this may be the way, and isn't too hard either with virtualbox. I guess AIR still has a leg up in that I can preview/run on windows

**Comment by ajwfrost** [comment]
From what I've seen, and the reports/information we've received, these issues are not actually 100% clear/reproducible across the board:
- I think the problem where Apple had been rejecting apps due to 'non-public APIs' may have gone away - I don't know whether that's because they've updated their scanners or not, but there were a number of initial reports of this but then some people have said they've submitted Windows-built applications that used the LLVM-based linker, and they were accepted into the App Store...
- The instabilities that can come from running a newly built (against stubs) app on older iOS versions is actually down to what APIs are used, there are some APIs that have moved from one library into another and that's something that the stub-based linking can't cope with. But those APIs are - I believe - ones that aren't called by the core AIR runtime, they happen when people use some third party frameworks that are included via ANEs. So it's not every app that will face this problem.

But in general, yes are finding that it's going to be impractical to keep the full Apple linker capabilities working on a Windows machine; you'd even find that the latest Xcode linker won't work when using the stub files on a Mac - we had to switch it to "ld-classic" (i.e. the linker from Xcode 14 and earlier). If you specify the platformsdk option on a Mac, and don't build against the stubs from the AIR SDK, then you can use the latest linker, which should really be the way forwards.

We had been working on a way to get Windows to do this build already. And with the next 51.0 beta release, we have got most of this in place, but the actual copying/triggering of the linker step is currently manual... basically the workflow would be:
1) set up the configuration so that ADT generates a folder with all the files needed for linking, plus a script file, in the specified location on your Windows machine
2) currently you'd then need to copy that over to a Mac, run the script, and copy back the executable file that results from it
3) then you can run ADT again, but with an additional argument specifying where the executable file is on your machine, and that will be used rather than the linker step, and will be signed and incorporated into the IPA

Obviously the next step in the plan is to make that seamless, by allowing you to configure a Mac machine that we can access via SSH, so that we can do that step 2 whilst ADT is still running, making it completely transparent to you. And from what I can tell, there shouldn't be any reasons why the Mac has to be local, it just has to be accessible across a network via SSH.

thanks

**Reply by mrfrasier** [reply]
@ajwfrost 
> * I think the problem where Apple had been rejecting apps due to 'non-public APIs' may have gone away - I don't know whether that's because they've updated their scanners or not, but there were a number of initial reports of this but then some people have said they've submitted Windows-built applications that used the LLVM-based linker, and they were accepted into the App Store...

I submitted a build today using 50.2.4.4 and got the error still

`ITMS-90338: Non-public API usage - The app contains one or more corrupted binaries. Rebuild the app and resubmit.. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/`

So definitely does still seem to be a thing. If i is something specific to a library, I'll leave my apm list here just in case it may help with data gathering on the issue

├──com.android.installreferrer@2.2.4
├──com.distriqt.AppleSignIn@2.3.0
├──com.distriqt.Application@7.1.0
├──com.distriqt.ApplicationRater@6.3.2
├──com.distriqt.Bolts@4.2.1
├──com.distriqt.CameraRollExtended@7.0.2
├──com.distriqt.IDFA@5.2.0
├──com.distriqt.Image@5.2.0
├──com.distriqt.InAppBilling@15.2.1
├──com.distriqt.ironsource.AdMob@2.0.4
├──com.distriqt.ironsource.AppLovin@11.10.2
├──com.distriqt.ironsource.FacebookAudience@6.16.1
├──com.distriqt.ironsource.UnityAds@4.8.0
├──com.distriqt.PushNotifications-FCM@14.1.0
├──com.distriqt.Scanner@5.1.0
└──com.distriqt.Share@7.6.0

plus any dependencies of course.

Also included but not part of APM is  com.digitalstrawberry.nativeExtensions.ANESounds

*edit* I removed ANESounds (since it seems playing sounds on Android isn't as big of an issue as it used to be) and resubmitted, but got the same result. Can only assume then it's one of the other ANEs. These are all Distriqt so maybe @marchbold has an idea.

**Reply by marchbold** [reply]
Do they give you a list of the private APIs that are being used? 

I'm fairly certain we don't use any private api's in our ANEs, we have carefully avoided them over the years in order to avoid these rejections.

**Reply by mrfrasier** [reply]
> Do they give you a list of the private APIs that are being used?

Of course not, this is Apple haha. That is literally the extent of the email aside from a generic link to their portal. I plan to sit down one day and systematically remove ANEs until it submits and see where that takes me. That day isn't today though haha

*edit* because it was easy, I did a build removing ALL the ironsource ANEs and adapters. Same issue, so it's something else (or the SDK? maybe I'll try an older release of AIR). I have notes from December where @ajwfrost said 50.2.2.2 would compile, but then I think we'd be right back to the issue of it crashing on iOS versions under 16

**Reply by ajwfrost** [reply]
I don't actually think there are any private APIs being used here, I think this is just an issue that's caused by the use of the LLVM linker. If you inspect the outputs of binaries created with LLVM vs with LD64, there are quite a lot of differences in the symbols that are included... which implies something not quite right with LLVM. Of course, the binaries still work/run fine, and it's probably not a big priority for the LLVM team.... hence the idea of linking these binaries "properly" on a mac, I would bet that there are no issues if you have all the same ANEs but link it there...

Currently we're working on the remote-linking thing, but hopefully it's possible to do this via a manual step now?

**Reply by mrfrasier** [reply]
> Currently we're working on the remote-linking thing, but hopefully it's possible to do this via a manual step now?

I'm all ears on how to try haha

**Reply by ajwfrost** [reply]
From https://airsdk.harman.com/api/versions/51.0.0.4/release-notes/Release_Notes_AIR_SDK_51.0.0.pdf

A new mechanism is being set up so that the “link” phase of building an IPA file can be carried out on a remote macOS machine, when developing on Windows. At this stage, there are three steps:
1. In the build configuration file (adt.cfg), there is a new setting that can be provided, “IPALinkFolder”. If specified, then rather than using/cleaning a temporary location, AIR will create a subfolder under the specified directory and will put all the necessary files for linking the application’s executable into this folder.
2. It also generates a script file (.sh). So once ADT has been run, and the subfolder has been created under the IPALinkFolder location, that subfolder can be zipped up and copied over to a macOS computer. An environment variable (AIR_SDK_HOME) needs to point to the root of an AIR SDK installation, and then the linker build script can be run to build the executable file.
3. Once the executable has been created on the macOS computer, it should be copied back onto the Windows PC. ADT can be run again as before, but this time providing an option “-use-linker-output” followed by the path to the executable file from step 2. This will cause ADT to skip the “link” phase and to just treat that file as if it were the output of linking.

We'd love to get some feedback on how this works in practice with more complicated applications..!

**Comment by mrfrasier** [comment]
So I guess somewhat similar to what Andrew suggested above, I now just share the project folder on my PC with the OSX vm. I still do all my building and testing on Windows, but when I need to submit to app store I switch over to the OSX VM and run the adt build command there.

Submitted to the App Store today without issue.
